ocaml/xapi/xapi_cluster.ml

```diff
@@ -116,84 +121,62 @@ let pool_create ~__context ~network ~cluster_stack ~token_timeout ~token_timeout
+    |> Xapi_stdext_monadic.Opt.unbox
   in
   let slave_cluster_hosts =
-    Db.Cluster.get_cluster_hosts ~__context ~self |> filter_on_option master_cluster_host
+    Db.Cluster.get_cluster_hosts ~__context ~self
+    |> List.filter ((<>) master_cluster_host)
   in
-  debug "Destroying cluster_hosts in pool";
-  (* First try to destroy each cluster_host - if we can do so safely then do *)
-  List.iter
-    (fun cluster_host ->
-      (* We need to run this code on the slave *)
-      (* We ignore failures here, we'll try a force_destroy after *)
-      log_and_ignore_exn (fun () ->
-        Helpers.call_api_functions ~__context (fun rpc session_id ->
-            Client.Client.Cluster_host.destroy ~rpc ~session_id ~self:cluster_host)
-      )
-    ) slave_cluster_hosts;
-  (* We expect destroy to have failed for some, we'll try to force destroy those *)
+  foreach_cluster_host ~__context ~self ~log:force
+    ~fn:Client.Client.Cluster_host.destroy
+    slave_cluster_hosts
+
+let pool_force_destroy ~__context ~self =
+  (* Set pool_autojoin:false and try to destroy slave cluster_hosts *)
+  pool_destroy_common ~__context ~self ~force:true;
+
   (* Note we include the master here, we should attempt to force destroy it *)
   let all_remaining_cluster_hosts =
     Db.Cluster.get_cluster_hosts ~__context ~self
   in
   (* Now try to force_destroy, keep track of any errors here *)
-  let exns = List.fold_left
-    (fun exns_so_far cluster_host ->
-      Helpers.call_api_functions ~__context (fun rpc session_id ->
-        try
-          Client.Client.Cluster_host.force_destroy ~rpc ~session_id ~self:cluster_host;
-          exns_so_far
-        with e ->
-          Backtrace.is_important e;
-          let uuid = Client.Client.Cluster_host.get_uuid ~rpc ~session_id ~self:cluster_host in
-          debug "Ignoring exception while trying to force destroy cluster host %s: %s" uuid (ExnHelper.string_of_exn e);
-          e :: exns_so_far
-      )
-    ) [] all_remaining_cluster_hosts
-    in
-
-    begin match exns with
-      | [] -> D.debug "Successfully destroyed all cluster_hosts in pool, now destroying cluster %s" (Ref.string_of self)
-      | e :: _ -> raise Api_errors.(Server_error (cluster_force_destroy_failed, [Ref.string_of self]))
-    end;
-
-    Helpers.call_api_functions ~__context (fun rpc session_id ->
-      Client.Client.Cluster.destroy ~rpc ~session_id ~self);
-    debug "Cluster_host.force_destroy was successful"
-
-(* Helper function; concurrency checks are done in implementation of Cluster.destroy and Cluster_host.destroy *)
+  debug "Ignoring exceptions while trying to force destroy cluster hosts.";
+    foreach_cluster_host ~__context ~self ~log:false
+      ~fn:Client.Client.Cluster_host.force_destroy
+      all_remaining_cluster_hosts;
+
+  (* Finally, any cluster_hosts we couldn't force destroy will be forgotten *)
+  begin match Db.Cluster_host.get_all ~__context with
```

Comment:
You probably don't need the match here, just call foreach with an empty list, it should be a no-op.